libraries/AP_Notify/OreoLED_I2C.cpp

```diff
@@ -0,0 +1,720 @@
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  OreoLED I2C driver
+*/
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/I2CDevice.h>
+#include <AP_Common/Semaphore.h>
+
+#include <AP_BoardConfig/AP_BoardConfig.h>
+#include "OreoLED_I2C.h"
+#include "AP_Notify.h"
+#include <utility>
+
+// OreoLEDs start at address 0x68 and add device number. So LED2 is at 0x6A
+#define OREOLED_BASE_I2C_ADDR           0x68
+
+#define OREOLED_BACKLEFT                0       // back left led instance number
+#define OREOLED_BACKRIGHT               1       // back right led instance number
+#define OREOLED_FRONTRIGHT              2       // front right led instance number
+#define OREOLED_FRONTLEFT               3       // front left led instance number
+#define PERIOD_SLOW                     800     // slow flash rate
+#define PERIOD_FAST                     500     // fast flash rate
+#define PERIOD_SUPER                    150     // super fast rate
+#define PO_ALTERNATE                    180     // 180 degree phase offset
+
+extern const AP_HAL::HAL& hal;
+
+// constructor
+OreoLED_I2C::OreoLED_I2C(uint8_t bus, uint8_t theme):
+    NotifyDevice(),
+    _bus(bus),
+    _oreo_theme(theme)
+{
+}
+
+//
+// Initialize the LEDs
+//
+bool OreoLED_I2C::init()
+{
+    // first look for led on external bus
+    _dev = std::move(hal.i2c_mgr->get_device(_bus, OREOLED_BASE_I2C_ADDR));
+    if (!_dev) {
+        return false;
+    }
+
+    // register timer
+    _dev->register_periodic_callback(1000, FUNCTOR_BIND_MEMBER(&OreoLED_I2C::update_timer, void));
+
+    _overall_health = true;
+
+    // return health
+    return true;
+}
+
+// UPDATE device according to timed_updated. Called at 50Hz
+void OreoLED_I2C::update()
+{
+
+    if (!_overall_health) {
+        return;    // don't go any further if LED driver reports unhealthy
+    }
+
+    if (slow_counter()) {
+        return;    // slow rate from 50hz to 10hz
+    }
+
+    sync_counter();                         // syncronizes LEDs every 10 seconds
+
+    if (mode_firmware_update()) {
+        return;    // don't go any further if the Pixhawk is in firmware update
+    }
+
+    if (mode_init()) {
+        return;    // don't go any further if the Pixhawk is initializing
+    }
+
+    if (mode_failsafe_radio()) {
+        return;    // don't go any further if the Pixhawk is is in radio failsafe
+    }
+
+    set_standard_colors();                  // set the rear LED standard colors as described above
```

Comment:
described where above? The comment is confusing.

Reply:
removed